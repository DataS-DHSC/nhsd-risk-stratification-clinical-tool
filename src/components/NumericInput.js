import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Fieldset, Input } from 'nhsuk-react-components';
import ConvertUnits from '../utils/convertUnits';

const NumericInput = ({
  field,
  prevData,
  required,
  groupName,
  changeCallback,
  setField,
}) => {
  const { label, ariaLabel, hint, hintAriaLabel, min, max, conversion } = field;

  const [localValue, setLocalValue] = useState(prevData[groupName]);

  const onChangeHandler = (e) => {
    setLocalValue(e.target.value);
    setField(groupName, e.target.value);
    changeCallback(groupName, e.target.value);
  };

  const onBlurHandler = (e) => {
    if (e.target.value) {
      setLocalValue(e.target.value);
      setField(groupName, Number(e.target.value).toString());
    }
  };

  const convertedUnits = ConvertUnits(localValue, conversion, 0);

  return (
    <Fieldset>
      <Fieldset.Legend className="hidden">{label}</Fieldset.Legend>

      <label
        className="nhsuk-heading-xs nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2"
        htmlFor={groupName}
        aria-label={`${ariaLabel}${required ? '' : ' (optional)'}`}
        id={`${field.label.toLowerCase().replace(/ /g, '-')}-radios-label`}
      >
        {`${label}${required ? '' : ' (optional)'}`}
      </label>
      <span
        className="nhsuk-hint"
        id={`${groupName}--hint`}
        aria-label={hintAriaLabel}
      >
        {hint}
      </span>
      <Input
        className="numeric-input nhsuk-input nhsuk-input--width-4"
        autoComplete="off"
        name={groupName}
        id={groupName}
        type="number"
        value={localValue}
        min={min}
        max={max}
        onChange={onChangeHandler}
        onBlur={onBlurHandler}
        required={required}
      />
      {conversion && localValue && (
        <div className="numeric-input-conversion">
          <span
            name={`${groupName}-conversion`}
            id={`${groupName}-conversion`}
            className="nhsuk-hint"
            aria-label={`(${convertedUnits.ariaLabel})`}
          >
            {`(${convertedUnits.label})`}
          </span>
        </div>
      )}
    </Fieldset>
  );
};

NumericInput.propTypes = {
  field: PropTypes.shape().isRequired,
  prevData: PropTypes.shape(),
  groupName: PropTypes.string.isRequired,
  required: PropTypes.bool,
  changeCallback: PropTypes.func.isRequired,
  setField: PropTypes.func.isRequired,
};

NumericInput.defaultProps = {
  prevData: {},
  required: false,
};

export default NumericInput;
