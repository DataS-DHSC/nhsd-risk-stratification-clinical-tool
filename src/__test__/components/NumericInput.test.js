import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import NumericInput from '../../components/NumericInput';

const defaultProps = {
  field: {
    label: 'test label',
    ariaLabel: 'test label aria',
    hint: 'test hint',
    hintAriaLabel: 'test hint aria',
    min: 0,
    max: 100,
  },
  required: true,
  prevData: { testField: '10' },
  groupName: 'testField',
  setField: jest.fn(),
  changeCallback: jest.fn(),
};

describe('NumericInput', () => {
  it('should match snapshot before and after conversion', () => {
    const props = {
      ...defaultProps,
      field: {
        label: 'test label',
        ariaLabel: 'test label aria',
        hint: 'test hint',
        hintAriaLabel: 'test hint aria',
        min: 0,
        max: 100,
        conversion: 'cm-to-ft-in',
      },
    };

    const { getByRole, asFragment } = render(<NumericInput {...props} />);
    expect(asFragment()).toMatchSnapshot();

    fireEvent.change(getByRole('spinbutton'), { target: { value: '55' } });

    expect(asFragment()).toMatchSnapshot();
  });

  it('sets state and calls callback on change', () => {
    const setField = jest.fn();
    const changeCallback = jest.fn();
    const props = {
      ...defaultProps,
      setField,
      changeCallback,
    };

    const { getByRole, asFragment } = render(<NumericInput {...props} />);
    expect(asFragment()).toMatchSnapshot();

    fireEvent.change(getByRole('spinbutton'), { target: { value: '55' } });

    expect(setField).toHaveBeenCalledWith('testField', '55');
    expect(changeCallback).toHaveBeenCalledWith('testField', '55');
    expect(asFragment()).toMatchSnapshot();
  });

  it('removes leading zeros onBlur', () => {
    const setField = jest.fn();
    const props = {
      ...defaultProps,
      setField,
    };

    const { getByRole } = render(<NumericInput {...props} />);

    fireEvent.blur(getByRole('spinbutton'), { target: { value: '055' } });

    expect(setField).toHaveBeenCalledWith('testField', '55');
  });
});
