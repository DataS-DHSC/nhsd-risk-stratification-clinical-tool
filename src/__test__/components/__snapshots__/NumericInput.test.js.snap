// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericInput sets state and calls callback on change 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <fieldset
      class="nhsuk-fieldset"
    >
      <legend
        class="nhsuk-fieldset__legend hidden"
      >
        test label
      </legend>
      <label
        aria-label="test label aria"
        class="nhsuk-heading-xs nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2"
        for="testField"
        id="test-label-radios-label"
      >
        test label
      </label>
      <span
        aria-label="test hint aria"
        class="nhsuk-hint"
        id="testField--hint"
      >
        test hint
      </span>
      <div
        class="nhsuk-form-group"
      >
        <input
          autocomplete="off"
          class="nhsuk-input numeric-input nhsuk-input nhsuk-input--width-4"
          id="testField"
          max="100"
          min="0"
          name="testField"
          required=""
          type="number"
          value="10"
        />
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;

exports[`NumericInput sets state and calls callback on change 2`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <fieldset
      class="nhsuk-fieldset"
    >
      <legend
        class="nhsuk-fieldset__legend hidden"
      >
        test label
      </legend>
      <label
        aria-label="test label aria"
        class="nhsuk-heading-xs nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2"
        for="testField"
        id="test-label-radios-label"
      >
        test label
      </label>
      <span
        aria-label="test hint aria"
        class="nhsuk-hint"
        id="testField--hint"
      >
        test hint
      </span>
      <div
        class="nhsuk-form-group"
      >
        <input
          autocomplete="off"
          class="nhsuk-input numeric-input nhsuk-input nhsuk-input--width-4"
          id="testField"
          max="100"
          min="0"
          name="testField"
          required=""
          type="number"
          value="55"
        />
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;

exports[`NumericInput should match snapshot before and after conversion 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <fieldset
      class="nhsuk-fieldset"
    >
      <legend
        class="nhsuk-fieldset__legend hidden"
      >
        test label
      </legend>
      <label
        aria-label="test label aria"
        class="nhsuk-heading-xs nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2"
        for="testField"
        id="test-label-radios-label"
      >
        test label
      </label>
      <span
        aria-label="test hint aria"
        class="nhsuk-hint"
        id="testField--hint"
      >
        test hint
      </span>
      <div
        class="nhsuk-form-group"
      >
        <input
          autocomplete="off"
          class="nhsuk-input numeric-input nhsuk-input nhsuk-input--width-4"
          id="testField"
          max="100"
          min="0"
          name="testField"
          required=""
          type="number"
          value="10"
        />
      </div>
      <div
        class="numeric-input-conversion"
      >
        <span
          aria-label="(0 feet 4 inches)"
          class="nhsuk-hint"
          id="testField-conversion"
          name="testField-conversion"
        >
          (0ft 4in)
        </span>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;

exports[`NumericInput should match snapshot before and after conversion 2`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <fieldset
      class="nhsuk-fieldset"
    >
      <legend
        class="nhsuk-fieldset__legend hidden"
      >
        test label
      </legend>
      <label
        aria-label="test label aria"
        class="nhsuk-heading-xs nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2"
        for="testField"
        id="test-label-radios-label"
      >
        test label
      </label>
      <span
        aria-label="test hint aria"
        class="nhsuk-hint"
        id="testField--hint"
      >
        test hint
      </span>
      <div
        class="nhsuk-form-group"
      >
        <input
          autocomplete="off"
          class="nhsuk-input numeric-input nhsuk-input nhsuk-input--width-4"
          id="testField"
          max="100"
          min="0"
          name="testField"
          required=""
          type="number"
          value="55"
        />
      </div>
      <div
        class="numeric-input-conversion"
      >
        <span
          aria-label="(1 feet 10 inches)"
          class="nhsuk-hint"
          id="testField-conversion"
          name="testField-conversion"
        >
          (1ft 10in)
        </span>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;
